### ISSUE TITLE
MAS Maneuver Planner

### ISSUE THREAD

**Issue by MOARdV** [MOARdV/AvionicsSystems#107]
Not a full MechJeb replacement, but a fairly simple design to automate some capabilities MAS already has.  The idea is that MAS has enough information to generate maneuver nodes.  I am not writing an autopilot, but generating nodes should be a simple piece of automation.

- [x] Change Apoapsis (easy)
- [x] Change Periapsis (easy)
- [x] Transfer to Target (easy, as long as target is in same SoI).
- [x] Return from Moon (easy)
- [x] Circularize At (not as easy)
- [x] Match Velocity at Closest Approach (maybe easy or not as easy)
- [x] Change Incllination
- [x] Match Planes

**Comment by MOARdV** [comment]
This is interesting - a lot of these equations are much simpler than I would have expected.  I looked at what MechJeb does for some of them, such as the Change Ap and Change Pe methods, and MJ does an iterative solver that does a binary search over a possible solution space.  However, for the most efficient cases (such as Change Pe when at Ap and Change Ap when at Pe), a direct equation provides the answer without having to do a search.

TL; DR - I think for the basic maneuvers that MAS can do, it is faster than MechJeb.

**Comment by MOARdV** [comment]
Last planned method is done.  I need to test ChangeInclination and MatchPlane more - I suspect I need to change the direction of the burn depending on whether I'm at AN or DN, and I don't do that currently.

**Comment by Alexustas** [comment]
I did a bit of testing the maneuver planner and here's some feedback:
1. The change of Apoapsis, Periapsis and Inclination looks good
2. The Circularization of the orbit with a small difference between Apoapsis and Periapsis also works well, but with a significant difference, the result is not very good. for example, I had Apopapsis 400 km and Periapsis 85 km and I produce a Circularization at an altitude of 100 km. the result of the MAS calculations give an orbit of 100,000 m to 84600 m, or something like that
3. Homman Transfer to the target vessel works well. Homman Transfer from Kerbin to Mun gave a strange result: Periapsis about 400000 Km and Inclination about 90 deg.
4. Match Velocity at Closest Approach works very incomprehensibly, the rel.inclination changes (in my case, from 0 to about 2.8 degrees), and the relative velocity after the maneuvering was 100+ m / s
([video](https://youtu.be/4jiId8y1JYw?t=9m4s))
5. Return from Moon - this transition gave me this kind of orbit
![screenshot57](https://user-images.githubusercontent.com/6713671/36140481-3bb19448-10c3-11e8-83ff-c94c85b39190.png)

**Comment by MOARdV** [comment]
Rocket science is hard. :)

**Comment by Alexustas** [comment]
> Homman Transfer to the target vessel works well.

as I discovered recently, it works well if I have a **circular** orbit. if the orbit is elliptical, for example 250 x 100 km, then the result of the calculation is rather strange

**Comment by MOARdV** [comment]
> as I discovered recently, it works well if I have a circular orbit. if the orbit is elliptical, for example 250 x 100 km, then the result of the calculation is rather strange

I believe this is also true of MechJeb.  The Maneuver Planner in MJ will warn if eccentricity of your orbit is too high.  I maybe can write a special mode for higher eccentricity, but that will require more computation.

**Comment by Alexustas** [comment]
OK. Then I'll try to display some kind of warning on the FMC. 
What do you think is the limit of eccentricity that is acceptable when using these features of the Maneuver Planner

**Comment by MOARdV** [comment]
> What do you think is the limit of eccentricity that is acceptable when using these features of the Maneuver Planner

I have not tested limits with this algorithm.  MechJeb will show a warning if ecc > 0.2.  Maybe that is a good start until I can test it.

**Comment by MOARdV** [comment]
Improvements will be under issue #119

**Comment by Alexustas** [comment]
**transfer.MatchPlane()** create a maneuver node on a Descending Node (with target) **only**, even if the next node is a Ascending Node
![screenshot85](https://user-images.githubusercontent.com/6713671/37866644-5d6060be-2faf-11e8-8f65-235cb40b0d80.png)

**Comment by MOARdV** [comment]
> create a maneuver node on a Descending Node (with target) only, even if the next node is a Ascending Node

MAS chooses the smallest ΔV node when matching planes.

**Comment by Alexustas** [comment]
OK got it

Video: https://youtu.be/FEsd2iFBXmo?t=6m56s

Here I have rel. inclination 1.55 and transfer.MatchPlane() create a maneuver after which the inclination should be 3.10 (twice as much) instead of 0, as if burn should be done in the opposite direction

**Comment by MOARdV** [comment]
Fixed relative inclination bug on [DropBox](https://www.dropbox.com/s/lgffr6rtdf0cc6b/AvionicsSystems-0.14.0b2.zip?dl=0).
